TFM: Fix Musca/AN521_nonsecure platforms to work with TFM building with or without BL2

ioannisg opened the pull request:
Currentlty the AN521 has a fixed offset for the BL2 header in the beginning of the non-secure image. This does not allow one to use AN521 with TFM without BL2 (which is a configurable option).

We fix this by utilizing ROM_START_OFFSET to create the placeholder for the BL2 header.

ioannisg committed:
modules: tfm: make BL2 header in non-secure image configurable

When we build Zephyr for a Non-Secure image, using TFM
as the Secure counterpart, we may or may not use BL2 (e.g.
MCUboot in TFM). If we build with BL2 we need to account
for a ROM offset before the start of the non-secure image
where TFM BL2 expects the BL2 header. This offset is not
needed when TFM is built without BL2. We use the existing
Kconfig options to determine whether we need a ROM offset
or not. We use the value of 0x400 for the BL2 header,
which is the default value used in TFM, but the option is
non-hidden, allowing the user to overwrite the default
value if needed.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: do not offset the image by default by 0x400.

We only need to offset the start of the non-secure image
by 0x400, if TFM is built with BL2 support. In this case
we use the ROM_START_OFFSET Kconfig switch  and set to
0x400. This instructs the linker to offset the beginning
of the ROM section by 0x400. In other words, we do not need
to statically move the start of the image by 0x400. This
fixes an issue that prevents from running Zephyr + TFM
without BL2 on MPS2 AN521.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

anangl commented on boards/arm/mps2_an521/mps2_an521_nonsecure.dts:
```diff
@@ -77,8 +77,8 @@
 
-	flash0: flash@100400 {
-		reg = <0x100400 0xDF00000>;
+	flash0: flash@100000 {
+		reg = <0x100000 0xDF00000>;
```
Together with this change you need to modify the way the non-secure image is signed because now `zephyr.bin` will contain the gap resulting from `ROM_START_OFFSET` at its beginning.
You need to add `--included-header` in this invocation of `imgtool.py`: https://github.com/zephyrproject-rtos/zephyr/blob/2cd556a689e36c5cf19336dcacafb41b900175e6/boards/arm/mps2_an521/CMakeLists.txt#L52-L63

ioannisg committed:
boards: arm: musca_b1: do not offset the image by default by 0x400.

We only need to offset the start of the non-secure image
by 0x400, if TFM is built with BL2 support. In this case
we use the ROM_START_OFFSET Kconfig switch  and set to
0x400. This instructs the linker to offset the beginning
of the ROM section by 0x400. In other words, we do not need
to statically move the start of the image by 0x400. This
fixes an issue that prevents from running Zephyr + TFM
without BL2 on Musca B1.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

anangl reviewed (approved):
LGTM

ioannisg commented:
@erwango could you check if this fix is required for the nucleo board the can build with TFM?

oyvindronningstad commented on modules/Kconfig.tfm:
```diff
@@ -56,4 +56,16 @@ config TFM_BL2_FALSE
+	default 0x400
+	help
+	  By default BL2 header size in TF-M is 0x400. ROM_START_OFFSET
+	  needs to be updated if TF-M switches to use a differend header
```
```suggestion
	  needs to be updated if TF-M switches to use a different header
```

ioannisg replied:
done thanks @oyvindronningstad

ioannisg commented on soc/arm/arm/mps2/soc.c:
```diff
@@ -56,7 +51,7 @@ void wakeup_cpu1(void)
 	*(uint32_t *)(SSE_200_SYSTEM_CTRL_INITSVTOR1) =
 		CONFIG_FLASH_BASE_ADDRESS +
-		BL2_HEADER_SIZE +
+		CONFIG_ROM_START_OFFSET +
```
@microbuilder @anangl @oyvindronningstad may I ask you to re-review the PR and in particular, these additional changes in the soc.c for musca and mps2?

AFAIR, @microbuilder added these for BL2 header, which, now is configurable in TFM and is reflected in ROM_START_OFFSET. So the initial reset vector table needs to be offset by 0x400 only if BL2 is built, when building with TFM. Otherwise not.

oyvindronningstad replied:
Can this be done with linker symbols instead?

ioannisg replied:
for example? _vector_start?

oyvindronningstad replied:
Yep

ioannisg replied:
done now, thanks @oyvindronningstad
